docs: add producer concepts article (#824)
asyncapi/website: pages/docs/concepts/producer.md

> ---

Reviewer: @nelsonmic  FYI, each new Docs page needs to have this title and weight metadata added! (sorry I forgot to tell you earlier)  This title is what is displayed in the **left-hand navigation bar** and the weight determines in what order to show the pages under each section.

Since we merged Thulie's **concept** page with a `weight: 2`, I made your page a `weight: 3`.

Author: Okay, thank you for the update

title: Producer
weight: 3
---

## What is a Producer? 

> An event indicates a state change or update that is triggered by a user's/device's action; this could be something like an item being placed in a shopping cart on an e-commerce website, clicking the subscribe button on a YouTube channel, or a temperature change in a sensor. A producer is a type of application that senses that state change and then publishes that event as a message.

Reviewer 2: ```suggestion
An event indicates a state change or update that is triggered by a user's/device's action; this could be something like an item being placed in a shopping cart on an e-commerce website, clicking the subscribe button on a YouTube channel, or a temperature change in a sensor. 

A producer is a type of application that senses that state change and then publishes that event as a message.
```

I just suggest to separate it in 2 paragraphs, as section is called `What is a procuder` and final definition is kinda hidden as last sentence in a paragraph, wdyt? or many leave as it is, but bold **A producer** ?

Reviewer 3: `this could be something like an item being placed in a shopping cart on an e-commerce website, clicking the subscribe button on a YouTube channel, or a temperature change in a sensor.`

This is a long sentence. Maybe you can try using bullet points? Something like this:

This could be something like:
* An item being placed in a shopping cart on an e-commerce website.
* Clicking the subscribe button on a YouTube channel.
* A temperature change in a sensor.

Author: Sweet, thank you! I'll make the changes right away

> An event indicates a state change or update which is triggered by a user's/device's action; this could be something like an item being placed in a shopping cart on an e-commerce website or clicking the subscribe button on a YouTube channel or a temperature change in the sensor. A producer is a type of application that senses that state change and then publishes that event as a message.

Reviewer: ```suggestion
An event indicates a state change or update that is triggered by a user's/device's action; this could be something like an item being placed in a shopping cart on an e-commerce website, clicking the subscribe button on a YouTube channel, or a temperature change in a sensor. A producer is a type of application that senses that state change and then publishes that event as a message.
```

## Why do we need Producers? 

> One of the core concepts of event-driven architecture is the publish/subscribe communication model. Producers are publishers in this model and the first logical layer responsible for publishing messages to the broker. This way, any consumer subscribed to that event can handle the published messages.

Reviewer 2: ```suggestion
One of the core concepts of event-driven architecture is the publish/subscribe communication model. Producers are publishers in this model and the first logical layer responsible for publishing to the broker messages that others can subscribe to receive.
```

maybe let us not mention subscribers here to much? to not complicate message. Especially that producer do not care who and if any subscriber is awaiting the message.

> One of the core concepts of an event driven architecture is the publish/subscribe communication model. Producers are essentially publishers in this model, they are the first logical layer in an event driven architecture, and are responsible for publishing messages to the broker so that any consumer subscribed to that event can handle it.

Reviewer: ```suggestion
One of the core concepts of event-driven architecture is the publish/subscribe communication model. Producers are publishers in this model and the first logical layer responsible for publishing messages to the broker. This way, any consumer subscribed to that event can handle the published messages.
```

The antecedent for `it` wasn't clear. Let me know if I got it right :)

Author: Yes, you got it right.

```mermaid
flowchart TD
    a[Producer]-- Message 1 --->d[(Broker)]
    d -- Message 1 --->g[Consumer]
```

> The diagram above depicts the communication between a **producer** publishing events to a specific channel in a **broker** and a consumer subscribed to the same channel.

Reviewer 2: I think would be good to not mention channel in this document, to not introduce to many different concepts, I think it is ok to just write about "broker" in general, without talking about channels/topics/routes. This should be a separate concept doc

Reviewer: hmmm.... Because the term `channel` is one of the easiest terms to understand, I do not think it's too overloaded to mention it on this page.  (if nelson had ALSO mentioned topics and routes I would agree that it is too much for this Doc, but he kept it high level by only mentioning channels.)

Terms like `broker` are the ones harder to understand for new audiences, so if anything I'd be worried about the word `broker` in this doc, not the word `channel`.  (Ironically, as a newer person to the spec... when reading this doc I understood it better **because** he included `channels` in his explanation. I am more familiar with channels than brokers ...as most folks are!... so it helps me follow along better.)

Also, we kinda have to be able to mention how all the pieces connect, and I think that's what this doc does. 

The improvement I would recommend is that once we add a NEW `/channel` doc, then we link to it here when mentioning `channels`.

In some cases, an entity can be both a producer publishing messages to a specific channel in the broker and a consumer subscribing to messages from a different channel in the broker.

```mermaid
flowchart LR
    a[Producer]-- Message 1 --->c[(Broker)]
    b[Producer]-- Message 2 --->c
    c -- Message 1 --> b
    c -- Message 2 ---> d[Consumer]
```

> In the diagram above, we see a producer publishing messages to a specific channel and a consumer subscribing to messages from that channel. We also have a second producer who publishes to one channel but also subscribes to messages from another.

Reviewer: ```suggestion
In the diagram above, we see a producer publishing messages to a specific channel and a consumer subscribing to messages from that channel. We also have a second producer who publishes to one channel but subscribes to messages from another.
```

> ## What is a Producer? 

Reviewer: <img width="500" alt="AsyncAPI diagram of a producer" src="https://user-images.githubusercontent.com/19964402/173460299-baf9f52d-1b83-4ab4-b1d9-39c14d08601c.png">

just adding this for visual context so folks can see the first draft of your diagram and your understanding of this spec term so far .. this is good learning for me too! :)

Author: Thank you @alequetzalli

Reviewer 2: - from a technical point of view it is too colorful. When you look at diagram as developer or architect, if components have different colors or shapes, it means they are different things. But here, we have producers and consumers, so components have specific role, so all `producer` shapes and `consumer` shapes should be the same, as the "label" already describes what they are
- `Channel` rename to `Broker` as `Broker` is a server, and it can have multiple channels/topics. You can actually have 2 diagrams here, first simplified, with producers, broker and consumers. Then more complex, which shows that some producers send message to one topic, and some to the other one, and then some consumers listen to messages only from one topic. Below you can see how indicate topics on arrow level + that it makes sense to have one component that has label `Producer/Consumer` to show that it is also possible
![Screenshot 2022-06-14 at 12 51 49](https://user-images.githubusercontent.com/6995927/173561044-b718dbad-b374-4087-b37a-7a5c80e7ef5e.png)
- please indicate flow by using arrows, that producer sends to broker, and broker sends to consumer

have a look at https://www.youtube.com/watch?v=dhdLxW0vDro that might help.

I hope that helps 🍻

Author: Thank you for your feedback @derberg, it definitely cleared up somethings

Author: Hello @derberg , again thanks for your feedback. 

I need you to help me clarify some things:
-  You said, "Some producers send message to one topic, and some to the other one". I don't quite get that statement. 

Also, my understanding of a producer is that when an event occurs it sends that message to the broker, now what I want to know is does a producer have a particular channel in the broker where it sends its own messages so that any consumer subscribed to that channel can handle it?

Reviewer 2: > Thank you for your feedback @derberg, it definitely cleared up somethings

no worries, happy to help. I think it is not so much different than with UI Design. Just map `Producer` components to buttons in the UI, you do not want every button in the I look differently.

> Some producers send message to one topic, and some to the other one

broker can have multiple channels/topics, each for different information. So there can be producer pushing messages to topic for, for example, info about `new user account created` like `user.created` and other producer pushing to topic for `user.deleted` a message that user was deleted.

this streetlight example -> https://github.com/asyncapi/spec/blob/v2.4.0/examples/streetlights-kafka.yml#L41-L72 there are 3 channels with `subscribe` operation. It means that this device described with this AsyncAPI document, sends messages to the broker, different messages to different channels, like a message that streetlight is turned on 

> Also, my understanding of a producer is that when an event occurs it sends that message to the broker, now what I want to know is does a producer have a particular channel in the broker where it sends its own messages so that any consumer subscribed to that channel can handle it?

`when an event occurs it sends that message to the broker` yup, and yes, message has to be sent to specific channel/topic, and any subscriber can subscribe to that topic and receive messages coming from a producer. 

does it help? sorry for late reply...holidays 😅

Author: Yes it does help... Thank you @derberg

> An event indicates a state change or update which is triggered by a users action; this could be something like an item being placed in a shopping cart on an e-commerce website or clicking the subscribe button on a YouTube channel. A producer is a type of application that senses that state change then publishes that event as a message to the channel. Ideally what is being sent to the channel by the producer is a message called the event notification, and not the event itself. 

Reviewer 2: ```suggestion
An event indicates a state change or update which is triggered by a user's/device's action; this could be something like an item being placed in a shopping cart on an e-commerce website or clicking the subscribe button on a YouTube channel or a temperature change in the sensor. A producer is a type of application that senses that state change and then publishes that event as a message.
```

I think this should be fine,
added `device` to make it clear it is not only about user action, it can be device, or just simple service action

I also suggested removal of `channel` as I'm not sure it is relevant + sometimes it is called topic or queue and in the end it doesn't matter where producer sends the message, more important is what it does. Better maybe add a sentence where you say: "for example producer sends a message to a topic in the message broker" - so it is clear it is example
